component/common/loki/client/queue_client.go

```diff
@@ -0,0 +1,601 @@
+type StoppableWriteTo interface {
+	agentWal.WriteTo
+	Stoppable
+	StopNow()
+}
+
+type MarkerHandler interface {
+	UpdateReceivedData(segmentId, dataCount int) // Data queued for sending
+	UpdateSentData(segmentId, dataCount int)     // Data which was sent or given up on sending
+	Stoppable
+}
+
+// queuedBatch is a batch specific to a tenant, that is considered ready to be sent.
+type queuedBatch struct {
+	TenantID string
+	Batch    *batch
+}
+
+// queue wraps a buffered channel and a routine that reads from it, sending batches of entries.
+type queue struct {
+	client *queueClient
+	q      chan queuedBatch
+	quit   chan struct{}
+	wg     sync.WaitGroup
+	logger log.Logger
+}
+
+func newQueue(client *queueClient, size int, logger log.Logger) *queue {
+	q := queue{
+		client: client,
+		q:      make(chan queuedBatch, size),
+		quit:   make(chan struct{}),
+		logger: logger,
+	}
+
+	q.wg.Add(1)
+	go q.run()
+
+	return &q
+}
+
+// enqueue is a blocking operation to add to the send queue a batch ready to be sent.
+func (q *queue) enqueue(qb queuedBatch) {
+	q.q <- qb
+}
+
+// enqueueWithCancel tries to enqueue a batch, giving up if the supplied context times deadlines
+// times out. If the batch is successfully enqueued, it returns true.
+func (q *queue) enqueueWithCancel(ctx context.Context, qb queuedBatch) bool {
+	select {
+	case <-ctx.Done():
+		return false
+	case q.q <- qb:
+	}
+	return true
+}
+
+func (q *queue) run() {
+	defer q.wg.Done()
+
+	for {
+		select {
+		case <-q.quit:
+			return
+		case qb := <-q.q:
+			// Since inside the actual send operation a context with time out is used, we should exceed that timeout
+			// instead of cancelling this send operations, since that batch has been taken out of the queue.
+			q.client.sendBatch(context.Background(), qb.TenantID, qb.Batch)
+			// mark segment data for that batch as sent, even if the send operation failed
+			qb.Batch.reportAsSentData(q.client.markerHandler)
+		}
+	}
+}
+
+// closeAndDrain stops gracefully the queue. The process first stops the main routine that reads batches to be sent,
+// to instead drain the queue and send those batches from this thread, exiting if the supplied context deadline
+// is exceeded. Also, if the underlying buffered channel is fully drain, this will exit promptly.
+func (q *queue) closeAndDrain(ctx context.Context) {
+	// defer main channel closing
+	defer close(q.q)
+
+	// first stop main routine, and wait for it to signal
+	close(q.quit)
+	q.wg.Wait()
+
+	// keep reading messages from sendQueue until all have been consumed, or timeout is exceeded
+	for {
+		select {
+		case qb := <-q.q:
+			// drain uses the same timeout, so if a timeout was applied to the parent context, it can cancel the underlying
+			// send operation preemptively.
+			q.client.sendBatch(ctx, qb.TenantID, qb.Batch)
+			// mark segment data for that batch as sent, even if the send operation failed
+			qb.Batch.reportAsSentData(q.client.markerHandler)
```

Comment:
If this is doing the same as the `run()` function, then I wonder if we should write a special function which just does `sendBatch` + `reportAsSentData`, and that function can then be used both by `run` and `closeAndDrain`. I'm worried about someone modifying the `run()` function and forgetting to modify this one.

Reply:
Agree!